src/lib/Interpreter.hs

```diff
@@ -277,24 +274,34 @@ matchUPats _ _ = error "mismatched lengths"
 
 -- XXX: It is a bit shady to unsafePerformIO here since this runs during typechecking.
 -- We could have a partial version of the interpreter that fails when any IO is to happen.
-indices :: EnvReader m => Type n -> m n [Atom n]
-indices ty = fmap fromListE $ flip runScopedT1 (mempty :: IxCache n) $ do
-    env <- unsafeGetEnv
-    Distinct <- getDistinct
-    ix <- simplifiedIxInstance ty
-    let IdxRepVal size = unsafePerformIO $ runInterpM env $ evalMethod ix simpleIxSize UnitVal
-    return $ ListE $ unsafePerformIO $ runInterpM env $
-      forM [0..size - 1] $ evalMethod ix simpleUnsafeFromOrdinal . IdxRepVal
-    where
-      evalMethod ix method x = case method ix of
-        Abs decls lam -> do
-          evalDecls decls $ evalExpr $ App (Lam lam) $ sinkFromTop x NE.:| []
+liftBuilderInterp
+  :: EnvReader m
+  => (forall l. (Emits l, DExt n l) => BuilderM l (Atom l))
+  -> m n (Atom n)
+liftBuilderInterp cont = do
+  block <- buildBlockSimplified cont
```

Comment:
I think we've been ambiguous about it. Certainly the interpreter doesn't do things like linearization and transposition, which are only done in simplify. In this case, I added the simplification to handle dictionary projection but maybe it's better to implement that in the interpreter directly. This can be a hot code path. I'll try that.